frontend/apps/reader/modules/readerlink.lua

```diff
@@ -91,6 +92,138 @@ function ReaderLink:init()
 
     -- delegate gesture listener to readerui, NOP our own
     self.ges_events = nil
+
+    -- Set up buttons for alternative external link handling methods
+    self._external_buttons["01_copy"] = function(this, link_url)
+        return {
+            text = _("Copy"),
+            callback = function()
+                UIManager:close(this.dialog)
+            end,
+        }
+    end
+
+    self._external_buttons["02_qrcode"] = function(this, link_url)
+        return {
+            text = _("Show QR code"),
+            callback = function()
+                UIManager:close(this.dialog)
+                UIManager:show(QRMessage:new{
+                        text = link_url,
+                        width = Device.screen:getWidth(),
+                        height = Device.screen:getHeight()
+                })
+            end
+        }
+    end
+    self._external_buttons["04_browser"] = function(this, link_url)
+        return {
+            text = _("Open in browser"),
+            callback = function()
+                UIManager:close(this.dialog)
+                Device:openLink(link_url)
+            end,
+            show_in_dialog_func = function()
+                if Device:canOpenLink() then
+                    return true
+                end
+            end
+        }
+    end
+    self._external_buttons["05_wiki_lookup"] = function(this, link_url)
+        return {
+            text = _("Read online"),
+            callback = function()
+                UIManager:nextTick(function()
+                        UIManager:close(this.dialog)
+                        local wiki_lang, wiki_page = link_url:match([[https?://([^%.]+).wikipedia.org/wiki/([^/]+)]])
+                        self.ui:handleEvent(Event:new("LookupWikipedia", wiki_page, true, false, true, wiki_lang))
+                end)
+            end,
+            show_in_dialog_func = function()
+                local wiki_lang, wiki_page = link_url:match([[https?://([^%.]+).wikipedia.org/wiki/([^/]+)]])
+                if wiki_lang and wiki_page then
+                    logger.dbg("Wikipedia link:", wiki_lang, wiki_page)
+                    text = T(_("Would you like to read this Wikipedia %1 article?\n\n%2\n"), wiki_lang:upper(), wiki_page:gsub("_", " "))
+                    return true, text
+                else
+                    return false
+                end
+            end
+        }
+    end
+    local find_epub_fullpath = function(link_url)
```

Comment:
This could be a module local function, instead of being fabricated on each :init() - although it would be a bit far from here were it is used (but it would make this section more readable).
Anyway, mention "wikipedia" in its name: `find_wikipedia_epub_fullpath()`

Or may be put all the logic and regexp in a single function:
`is_wiki_page(link_uri, check_if_epub_exists) => wiki_page, wiki_lang, epub_fullpath`
which may make this section smaller ?

Reply:
Great suggestion, thanks!